File: sevenkplus/normal/NA/112.py
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;
typedef pair<int, int> PII;

#define fi first
#define se second
#define mp make_pair
#define pb push_back

#define N 200010
#define P 1000000007

int n;
int v[N];
int a[N];
vector<int> b[N];
int q[N], L;
//int f[N], g[N];

void ff(int x) {
	v[x] = 2;
	for (vector<int>::iterator i = b[x].begin(); i != b[x].end(); i ++)
		if (v[*i] != 2) ff(*i);
	q[L++] = x;
}

int main() {
	cin >> n;
	memset(a, -1, sizeof a);
	for (int i = 0; i < n; i ++) {
		int x, y;
		cin >> x >> y;
		x--; y--;
		a[x] = y;
		b[y].pb(x);
	}
	
	int S = 1;
	for (int i = 0; i < n*2; i ++)
		if (!v[i]) {
			int j = i;
			while (!v[j]) {
				v[j] = 1;
				if (a[j] == -1) break;
				j = a[j];
			}
			L = 0;
			ff(j);
			if (a[j] != -1) {
				if (a[j] != j) S = (ll)S*2%P;
				continue;
			}
			S = (ll)S*L%P;
		}
	cout << S << endl;
	return 0;
}

File: mnbvmar/normal/NA/112.cpp
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>

using namespace std;
#define PB push_back
#define MP make_pair
#define st first
#define nd second
#define ALL(x) (x).begin(), (x).end()
#define SZ(x) ((int)(x).size())

#define unordered_map __fast_unordered_map
template<class Key, class Value, class Hash = std::hash<Key>>
using unordered_map = __gnu_pbds::gp_hash_table<Key, Value, Hash>;

template<class C> void mini(C &a4, C b4) { a4 = min(a4, b4); }
template<class C> void maxi(C &a4, C b4) { a4 = max(a4, b4); }

template<class TH> void _dbg(const char *sdbg, TH h){ cerr<<sdbg<<'='<<h<<endl; }
template<class TH, class... TA> void _dbg(const char *sdbg, TH h, TA... a) {
  while(*sdbg!=',')cerr<<*sdbg++;
  cerr<<'='<<h<<','; _dbg(sdbg+1, a...);
}

template<class T> ostream &operator<<(ostream& os, vector<T> V) {
  os << "["; for (auto vv : V) os << vv << ","; return os << "]";
}
template<class L, class R> ostream &operator<<(ostream &os, pair<L,R> P) {
  return os << "(" << P.st << "," << P.nd << ")";
}

#ifdef LOCAL
#define debug(...) _dbg(#__VA_ARGS__, __VA_ARGS__)
#else
#define debug(...) (__VA_ARGS__)
#define cerr if(0)cout
#endif

using ll = long long;
using vi = vector<int>;
using pii = pair<int, int>;
using pll = pair<ll, ll>;
using vll = vector<ll>;
using ld = long double;


int32_t main() {
  ll n;
  ll m, k;

  scanf("%lld%lld%lld", &n, &m, &k);

  vll locs(m);
  for (ll &x : locs) { scanf("%lld", &x); }

  ll nops = 0;
  ll ptr = 0;

  auto GetLoc = [&](ll idx) {
    return locs[idx] - ptr - 1;
  };

  while (ptr < m) {
    const ll page = GetLoc(ptr) / k;
    ll x = ptr;
    while (x < m && GetLoc(x) / k == page) { ++x; }
    ptr = x;
    ++nops;
  }

  printf("%lld\n", nops);
}